### PR TITLE
Resolve signed-unsigned comparison issues in TypeSafeIndex

### LINKED ISSUE #13482: Comparison operators in TypeSafeIndex trigger GCC 9 sign-compare warning
For example, from https://drake-jenkins.csail.mit.edu/job/linux-focal-unprovisioned-gcc-bazel-experimental-debug/1/:
```
ERROR: manipulation/kuka_iiwa/BUILD.bazel:35:1: Couldn't build file manipulation/kuka_iiwa/_objs/iiwa_command_receiver/iiwa_command_receiver.pic.o: C++ compilation of rule '//manipulation/kuka_iiwa:iiwa_command_receiver' failed (Exit 1) gcc failed: error executing command /usr/bin/gcc -U_FORTIFY_SOURCE -fstack-protector -Wall -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer -g '-std=c++0x' -MD -MF ... (remaining 205 argument(s) skipped)
In file included from bazel-out/k8-dbg/bin/systems/framework/_virtual_includes/framework_common/drake/systems/framework/framework_common.h:13,
                 from bazel-out/k8-dbg/bin/systems/framework/_virtual_includes/continuous_state/drake/systems/framework/continuous_state.h:15,
                 from bazel-out/k8-dbg/bin/systems/framework/_virtual_includes/leaf_system/drake/systems/framework/leaf_system.h:22,
                 from bazel-out/k8-dbg/bin/manipulation/kuka_iiwa/_virtual_includes/iiwa_command_receiver/drake/manipulation/kuka_iiwa/iiwa_command_receiver.h:12,
                 from manipulation/kuka_iiwa/iiwa_command_receiver.cc:1:
bazel-out/k8-dbg/bin/common/_virtual_includes/type_safe_index/drake/common/type_safe_index.h: In instantiation of 'typename std::enable_if<(std::is_integral<U>::value && std::is_unsigned<U>::value), bool>::type drake::TypeSafeIndex<Tag>::operator==(const U&) const [with U = long unsigned int; Tag = drake::systems::DiscreteStateTag; typename std::enable_if<(std::is_integral<U>::value && std::is_unsigned<U>::value), bool>::type = bool]':
bazel-out/k8-dbg/bin/systems/framework/_virtual_includes/system_base/drake/systems/framework/system_base.h:905:5:   required from here
bazel-out/k8-dbg/bin/common/_virtual_includes/type_safe_index/drake/common/type_safe_index.h:351:18: error: comparison of integer expressions of different signedness: 'const long unsigned int' and 'int' [-Werror=sign-compare]
  351 |     return value <= std::numeric_limits<int>::max() &&
      |            ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
bazel-out/k8-dbg/bin/common/_virtual_includes/type_safe_index/drake/common/type_safe_index.h: In instantiation of 'typename std::enable_if<(std::is_integral<U>::value && std::is_unsigned<U>::value), bool>::type drake::TypeSafeIndex<Tag>::operator==(const U&) const [with U = long unsigned int; Tag = drake::systems::AbstractStateTag; typename std::enable_if<(std::is_integral<U>::value && std::is_unsigned<U>::value), bool>::type = bool]':
bazel-out/k8-dbg/bin/systems/framework/_virtual_includes/system_base/drake/systems/framework/system_base.h:919:5:   required from here
bazel-out/k8-dbg/bin/common/_virtual_includes/type_safe_index/drake/common/type_safe_index.h:351:18: error: comparison of integer expressions of different signedness: 'const long unsigned int' and 'int' [-Werror=sign-compare]
bazel-out/k8-dbg/bin/common/_virtual_includes/type_safe_index/drake/common/type_safe_index.h: In instantiation of 'typename std::enable_if<(std::is_integral<U>::value && std::is_unsigned<U>::value), bool>::type drake::TypeSafeIndex<Tag>::operator==(const U&) const [with U = long unsigned int; Tag = drake::systems::NumericParameterTag; typename std::enable_if<(std::is_integral<U>::value && std::is_unsigned<U>::value), bool>::type = bool]':
bazel-out/k8-dbg/bin/systems/framework/_virtual_includes/system_base/drake/systems/framework/system_base.h:931:5:   required from here
bazel-out/k8-dbg/bin/common/_virtual_includes/type_safe_index/drake/common/type_safe_index.h:351:18: error: comparison of integer expressions of different signedness: 'const long unsigned int' and 'int' [-Werror=sign-compare]
bazel-out/k8-dbg/bin/common/_virtual_includes/type_safe_index/drake/common/type_safe_index.h: In instantiation of 'typename std::enable_if<(std::is_integral<U>::value && std::is_unsigned<U>::value), bool>::type drake::TypeSafeIndex<Tag>::operator==(const U&) const [with U = long unsigned int; Tag = drake::systems::AbstractParameterTag; typename std::enable_if<(std::is_integral<U>::value && std::is_unsigned<U>::value), bool>::type = bool]':
bazel-out/k8-dbg/bin/systems/framework/_virtual_includes/system_base/drake/systems/framework/system_base.h:945:5:   required from here
bazel-out/k8-dbg/bin/common/_virtual_includes/type_safe_index/drake/common/type_safe_index.h:351:18: error: comparison of integer expressions of different signedness: 'const long unsigned int' and 'int' [-Werror=sign-compare]
bazel-out/k8-dbg/bin/common/_virtual_includes/type_safe_index/drake/common/type_safe_index.h: In instantiation of 'typename std::enable_if<(std::is_integral<U>::value && std::is_unsigned<U>::value), bool>::type drake::TypeSafeIndex<Tag>::operator<(const U&) const [with U = long unsigned int; Tag = drake::systems::DiscreteStateTag; typename std::enable_if<(std::is_integral<U>::value && std::is_unsigned<U>::value), bool>::type = bool]':
bazel-out/k8-dbg/bin/systems/framework/_virtual_includes/system_base/drake/systems/framework/system_base.h:1166:5:   required from here
bazel-out/k8-dbg/bin/common/_virtual_includes/type_safe_index/drake/common/type_safe_index.h:395:18: error: comparison of integer expressions of different signedness: 'const long unsigned int' and 'int' [-Werror=sign-compare]
  395 |     return value > std::numeric_limits<int>::max() ||
      |            ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
bazel-out/k8-dbg/bin/common/_virtual_includes/type_safe_index/drake/common/type_safe_index.h: In instantiation of 'typename std::enable_if<(std::is_integral<U>::value && std::is_unsigned<U>::value), bool>::type drake::TypeSafeIndex<Tag>::operator<(const U&) const [with U = long unsigned int; Tag = drake::systems::AbstractStateTag; typename std::enable_if<(std::is_integral<U>::value && std::is_unsigned<U>::value), bool>::type = bool]':
bazel-out/k8-dbg/bin/systems/framework/_virtual_includes/system_base/drake/systems/framework/system_base.h:1172:5:   required from here
bazel-out/k8-dbg/bin/common/_virtual_includes/type_safe_index/drake/common/type_safe_index.h:395:18: error: comparison of integer expressions of different signedness: 'const long unsigned int' and 'int' [-Werror=sign-compare]
bazel-out/k8-dbg/bin/common/_virtual_includes/type_safe_index/drake/common/type_safe_index.h: In instantiation of 'typename std::enable_if<(std::is_integral<U>::value && std::is_unsigned<U>::value), bool>::type drake::TypeSafeIndex<Tag>::operator<(const U&) const [with U = long unsigned int; Tag = drake::systems::NumericParameterTag; typename std::enable_if<(std::is_integral<U>::value && std::is_unsigned<U>::value), bool>::type = bool]':
bazel-out/k8-dbg/bin/systems/framework/_virtual_includes/system_base/drake/systems/framework/system_base.h:1178:5:   required from here
bazel-out/k8-dbg/bin/common/_virtual_includes/type_safe_index/drake/common/type_safe_index.h:395:18: error: comparison of integer expressions of different signedness: 'const long unsigned int' and 'int' [-Werror=sign-compare]
bazel-out/k8-dbg/bin/common/_virtual_includes/type_safe_index/drake/common/type_safe_index.h: In instantiation of 'typename std::enable_if<(std::is_integral<U>::value && std::is_unsigned<U>::value), bool>::type drake::TypeSafeIndex<Tag>::operator<(const U&) const [with U = long unsigned int; Tag = drake::systems::AbstractParameterTag; typename std::enable_if<(std::is_integral<U>::value && std::is_unsigned<U>::value), bool>::type = bool]':
bazel-out/k8-dbg/bin/systems/framework/_virtual_includes/system_base/drake/systems/framework/system_base.h:1184:5:   required from here
bazel-out/k8-dbg/bin/common/_virtual_includes/type_safe_index/drake/common/type_safe_index.h:395:18: error: comparison of integer expressions of different signedness: 'const long unsigned int' and 'int' [-Werror=sign-compare]
```

Relates #13102. FYI @jwnimmer-tri @SeanCurtis-TRI.

### PR DESCRIPTION
**Pull request by SeanCurtis-TRI**
When comparing an unsigned integral value with a TypeSafeIndex, we needed to account for the possibility that the unsigned integral value could contain larger values than the TypeSafeIndex. This was done by comparing the unsigned integer with the maximum int (the largest type safe
index value). This is intrinsically a signed-unsigned comparision which is problematic.

Instead, we cast the maximum int to the unsigned integral type. If the unsigned type is larger (more bits), the relationship will be fully preserved and unsigned values that are larger than can be represented will be detected.

If the unsigned type has fewer bits, the maximum int (01111.1111) will be truncated to a number consisting of all 1s, the maximum unsigned value of that number of bits. We are guaranteed that the unsigned _value_ must be less than or equal to that truncated value which means its less than or
equal to the full maximum index.

In other words:

value <= static_cast<U>(kMaxIndex) is always *true* if
    sizeof(U) < sizeof(int).

resolves #13482

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/13486)

### CODE REVIEW

**Comment by SeanCurtis-TRI** [comment]
@drake-jenkins-bot linux-focal-unprovisioned-gcc-bazel-experimental-debug please

**Comment by jamiesnape** [comment]
Lint is unhappy, but looks like the Focal build is going to be good otherwise.
